### PR TITLE
Improved mnist training with tracking loss and params, early stoping and saving best network

### LINKED ISSUE #376: Improve MNIST training test
Add validation after each training epoch (loss, accuracy).
Add early stopping logic.
Add Tensorboard loss plots.

After training, calculate accuracy on test set.

### PR DESCRIPTION
**Pull request by ndrakulicTT**
Test for full training that writes to tensorboard
Test for gradients after one pass of neural network

Added utils for:
- Handling params and gradients of network
- Generators for training and test loop
- class earlyStoping

Closes #376

### CODE REVIEW

File: forge/test/mlir/mnist/training/test_training.py [hunk 1]
```diff
@@ -93,3 +91,139 @@ def test_mnist_training():
...
+# That sets relu threshold to bfloat16 tensor.
+# And in file forge/forge/compile.py::compile_main forced bfloat 16 should be added compiler_cfg.default_df_override = DataFormat.Float16_b
+@pytest.mark.skip(reason="Need to be tested with bfloat16 and takes around 10 minutes to run")
+def test_forge_vs_torch():
+    torch.manual_seed(0)
+
+    batch_size = 64
+    learning_rate = 1e-2
+    epochs = 10
+    verobse = True
```
Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/training/test_training.py:152]: typo :)

File: forge/test/mlir/mnist/utils.py [hunk 2]
```diff
@@ -48,13 +72,81 @@ def load_dataset(batch_size):
...
+    for name in named_params:
+        writer.add_histogram(name, named_params[name].flatten().float(), step)
+
+
+def train_loop(dataloader, model, loss_fn, optimizer, batch_size, named_params, isTT=False, verbose=False):
+    size = len(dataloader.dataset)
+    for batch, (X, y) in enumerate(dataloader):
+        optimizer.zero_grad()
+        pred = model(X)
+        pred = pred[0] if isTT else pred
```
Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/utils.py:115]: use standard notation (with _) for isTT

File: forge/test/mlir/mnist/utils.py [hunk 3]
```diff
@@ -14,32 +14,56 @@
...
 
+class EarlyStopping:
+    def __init__(self, patience=3):
+        self.patience = patience
+        self.counter = 0
+        self.best_score = None
+        self.early_stop = False
 
-def load_tb_writer():
+    def __call__(self, val_metric):
```
Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/utils.py:39]: Sometimes you want to save model with largest val metric (accuracy) and sometimes with lowest (loss). This need to be factored in.

Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/utils.py:39]: Instead of __call__, I would use 'is_current_best', it was confusing for me.

File: forge/test/mlir/mnist/utils.py [hunk 4]
```diff
@@ -48,13 +72,81 @@ def load_dataset(batch_size):
...
+
+
+def train_loop(dataloader, model, loss_fn, optimizer, batch_size, named_params, isTT=False, verbose=False):
+    size = len(dataloader.dataset)
+    for batch, (X, y) in enumerate(dataloader):
+        optimizer.zero_grad()
+        pred = model(X)
+        pred = pred[0] if isTT else pred
+
+        y = nn.functional.one_hot(y.long(), num_classes=10).to(pred.dtype)
```
Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/utils.py:117]: why do we need y.long()?

File: forge/test/mlir/mnist/utils.py [hunk 5]
```diff
@@ -48,13 +72,81 @@ def load_dataset(batch_size):
...
+    with torch.no_grad():
+        for X, y in dataloader:
+            pred = model(X)
+            pred = pred[0] if isTT else pred
+            y = nn.functional.one_hot(y.long(), num_classes=10).to(pred.dtype)
+            test_loss += loss_fn(pred, y).item()
+            correct += (pred.argmax(1) == y.argmax(1)).type(torch.float).sum().item()
+
+    test_loss /= size
+    correct /= size
```
Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/utils.py:147]: nit: I would call this 'accuracy'

Comment (vladimirjovanovicTT) [reply forge/test/mlir/mnist/utils.py:147]: also, validation_loss

File: forge/test/mlir/mnist/training/test_training.py [hunk 6]
```diff
@@ -93,3 +91,139 @@ def test_mnist_training():
...
+                forge_writer.add_scalar("validation_acc", forge_val_acc, step)
+
+                torch_writer.flush()
+                forge_writer.flush()
+
+        if verobse:
+            print(f"Epoch {i} took {time.time() - start_time} seconds")
+
+        if earlyStop(forge_val_acc):
+            torch.save(torch_model.state_dict(), "runs/models/torch_model.pth")
```
[the remark this answers is not in the document]
Comment (vladimirjovanovicTT) [reply forge/test/mlir/mnist/training/test_training.py:225]: also, we can keep track of the current best model (optionally its accuracy/validation_loss) and load it at the end of the training so we can run a sample inference, or calculate accuracy on a test set.

File: forge/test/mlir/mnist/training/test_training.py [hunk 7]
```diff
@@ -93,3 +91,139 @@ def test_mnist_training():
...
+    # Compare gradients for each parameter
+    for name in reversed(list(torch_grads.keys())):
+        assert compare_with_golden(torch_grads[name], forge_grads[name])
+
+
+# For bfloat16, the following line should be added to the test_forge_vs_torch function:
+# In file forge/forge/op/eval/forge/eltwise_unary.py:418 should be replaced with: threshold_tensor = ac.tensor(torch.zeros(shape, dtype=torch.bfloat16) + threshold)
+# That sets relu threshold to bfloat16 tensor.
+# And in file forge/forge/compile.py::compile_main forced bfloat 16 should be added compiler_cfg.default_df_override = DataFormat.Float16_b
+@pytest.mark.skip(reason="Need to be tested with bfloat16 and takes around 10 minutes to run")
```
Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/training/test_training.py:145]: @pilkicTT @nvukobratTT we plan to move this out of forge-fe and to a new forge-training repo, which we will use to do training experiments/runs. How does this sound to you?

File: forge/test/mlir/mnist/training/test_training.py [hunk 8]
```diff
@@ -28,11 +29,8 @@ def test_mnist_training():
     # Load dataset
     test_loader, train_loader = load_dataset(batch_size)
 
-    # Load TensorBoard writer (for logging)
-    writer = load_tb_writer()
-
     # Define model and instruct it to compile and run on TT device
-    framework_model = MNISTLinear()
+    framework_model = MNISTLinear(bias=False)
```
Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/training/test_training.py:33]: Just add a comment here why bias needs to be False

File: forge/test/mlir/mnist/utils.py [hunk 9]
```diff
@@ -14,32 +15,73 @@
 
 # Model definition
 class MNISTLinear(nn.Module):
-    def __init__(self, input_size=784, output_size=10, hidden_size=256):
+    def __init__(self, input_size=784, output_size=10, hidden_size=512, bias=True, dtype=torch.float32):
```
Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/utils.py:18]: add comment why hidden_size is changed (ttnn matmul doesnt work for dims)

File: forge/test/mlir/mnist/utils.py [hunk 10]
```diff
@@ -14,32 +15,73 @@
...
+        self.patience = patience
+        self.counter = 0
+        self.best_score = None
+        self.early_stop = False
+        self.is_current_best = False
+        self.best_model = None
 
-def load_tb_writer():
+    def step(self, val_metric, model_id):
+        if self.best_score is None:
```
Comment (vladimirjovanovicTT) [inline forge/test/mlir/mnist/utils.py:48]: you can merge 'if is None' and 'if is best' into one

**Review by vladimirjovanovicTT — APPROVED** [review]
